torch/fx/experimental/symbolic_shapes.py

```diff
@@ -329,28 +329,62 @@ def _get_key(self):
         """
         return (len(self.replacements), len(self.divisible))
 
-    # NB: This is only called for input symbolic sizes; intermediate symbolic
-    # sizes are allocated via a different mechanism
-    def create_symint(self, name, val):
-        assert val >= 0
+    def create_symbolic_sizes_strides(self, ex: torch.Tensor):
+        size = [self[i] for i in ex.size()]
+        stride = [None] * len(size)
+        for i, val in enumerate(ex.stride()):
+            if val in (0, 1):
+                stride[i] = sympy.Integer(val)
+        while any(x is None for x in stride):
+            candidates = {
+                ex.size(i) * ex.stride()[i]: size[i] * stride[i]
+                for i in range(len(size))
+                if stride[i] is not None and ex.stride()[i] >= 0
+            }
+            # iterate over unbound strides in sorted order
+            val_list = sorted(
+                [(ex.stride()[i], i) for i in range(len(stride)) if stride[i] is None]
+            )
+            for _, i in val_list:
+                if stride[i] is None and ex.stride()[i] in candidates:
+                    stride[i] = candidates[ex.stride()[i]]
+                    candidates[ex.size(i) * ex.stride()[i]] = size[i] * stride[i]
+            if any(x is None for x in stride):
+                # bind the smallest unbound stride to a new variable
+                val, i = sorted(
+                    [
+                        (ex.stride()[i], i)
+                        for i in range(len(stride))
+                        if stride[i] is None
+                    ]
+                )[0]
+                stride[i] = self[val]
+
+        return [self.create_symintnode(i) for i in size], [self.create_symintnode(i) for i in stride]
+
+    def create_symintnode(self, expr: Union["sympy.Expr", int]):
+        if isinstance(expr, int):
+            return expr
+        py_sym_int = PySymInt(expr, self)
+        cpp_sym_int = torch.SymIntNode.new_symint(py_sym_int)  # type: ignore[attr-defined]
+        return cpp_sym_int
+
+    def __getitem__(self, val: int) -> "sympy.Expr":
         if not HAS_SYMPY:
             raise RuntimeError("Need sympy installed to create symbolic shapes")
-
-        # TODO: Put 0/1 specialization in guards
-        if val == 0 or val == 1:
-            return val
+        if val < 0:
+            # all variables are positive
```

Comment:
This is only for creating symbolic variables.

Reply:
Would it be clearer to say "all sympy base variables must be positive" ?

Reply:
just curious - is this just to establish some nice invariants? (all "base" values of a sympy expression are positive). Or does something break if we allow negative valued symbols.

Reply:
oh, hopefully answered my own question - it looks like in `_maybe_evaluate_static` ([here](https://github.com/pytorch/pytorch/blob/3356d0385fd0f3f0f6ce2d8c681a40fd110c7848/torch/fx/experimental/symbolic_shapes.py#L394)), we can take an expression like `Eq(symbol, 0)` and statically evaluate it to False without introducing guards, since we know that all of our symbols are positive?